### ISSUE TITLE
Meep Fails to Build (Ubuntu 18.04)

### ISSUE THREAD

**Issue by bcerjan** [NanoComp/meep#1102]
I'm trying to build Meep on Ubuntu 18.04 using the automake tools. In particular, I'm running configure (from `autogen.sh`) as follows:
`./configure --enable-maintainer-mode "$@" \
 CC=emcc CXX=em++ \
 --without-libctl \
 CPPFLAGS="-I/path/to/libctl_and_guile/headers" \
 --without-python --enable-shared \
 --enable-portable-binary --without-gcc-arch`

This step works fine, but when I try to `make` I get the following errors:
1. `meepgeom.cpp:282:20: error: no viable overloaded '='
  default_material = (void *)md;
  ctlgeom-types.h:15:16: note: candidate function (the implicit copy assignment operator) not viable: cannot convert argument of incomplete type 'void *' to 'const material_type_struct' for 1st argument`

2. `meepgeom.cpp:395:3: error: use of undeclared identifier 'geom_fix_object_list'
  geom_fix_object_list(geometry);`

3. (repeated several times) `meepgeom.cpp:523:7: error: cannot cast from type 'material_type' (aka 'material_type_struct') to pointer type 'meep_geom::material_type' (aka 'meep_geom::material_data *')
      (material_type)material_of_unshifted_point_in_tree_inobject(p, restricted_tree, &inobject);`

4. (several of these, for `is_variable`, `is_file`, `is_medium`) `meepgeom.cpp:667:14: error: no matching function for call to 'is_variable'
  if ((o1 && is_variable(o1->material)) || (o2 && is_variable(o2->material)) ||
             ^~~~~~~~~~~
meepgeom.cpp:245:6: note: candidate function not viable: no known conversion from 'const material_type' (aka 'const material_type_struct') to 'meep_geom::material_type' (aka 'meep_geom::material_data *') for 1st argument
bool is_variable(material_type mt) { return (mt->which_subclass == material_data::MATERIAL_USER); }`

I'm using the emscripten versions of gcc and g++ to compile to WebAssembly (so I really just need a cross-platform binary version of the meep library). However, when I try using `CC=gcc` and `CXX=g++` it fails with the same errors, which indicates that it's not just a difference with how emcc and gcc behave. It appears to be a namespace-scoping issue as the gcc compiler gives hints like:
`meepgeom.cpp:1047:50: error: no matching function for call to ‘is_medium(material_type&, meep_geom::medium_struct**)’
     if (is_medium(geometry.items[i].material, &mm))
                                                  ^
meepgeom.cpp:251:6: note: candidate: bool      meep_geom::is_medium(meep_geom::material_type, meep_geom::medium_struct**)
 bool is_medium(material_type md, medium_struct **m)`

Is there a standard flag I need to set to tell the compiler to use a particular version of C++ (or C)? Or is there some set of changes I need to make throughout the file to tell the compiler where to find all the functions it needs?

**Comment by danielwboyce** [comment]
I noticed in your `./configure` sequence you're passing a `--without-libctl` flag. `libctl` is one of `meep`'s core dependencies so `meep` will absolutely not compile correctly without it. A lot of the error messages seems to be related to parts of `meep` that are handled (at least in part) by `libctl`, as well. See if compiling with `libctl` helps, and maybe make sure [you're handling the other dependencies and the build sequence correctly](https://meep.readthedocs.io/en/latest/Build_From_Source/). As for compiling to WebAssembly, I'm not sure who all knows much about it on this repository... But compiling to support WebAssembly is probably talked about in WebAssembly-related forums, so maybe try there.

**Comment by bcerjan** [comment]
When I try to compile `--with-libctl=/path/to/libctl` I get an error: `configure: error: linking to guile failed` but I'm passing the location of guile via `LDFLAGS` (unless this is not how I should be telling it where to find guile).

If I try to build libctl `--without-guile` (I know that's a separate project) it fails at ` *** No rule to make target 'geom-ctl-io.c', needed by 'all'.  Stop.`

Perhaps obviously, I'm trying to use as few extra libraries/tools as necessary since I believe I need to compile them all individually using emcc/em++ and it'd be easier not to have to do that (and I only need the C++ interface for `meep`).

**Comment by doswellf** [comment]
Before trying to build meep I built and installed hdf5-parallel and h5py. 
I will describe the whole process in a longer post or on my github pages.

As for meep cloned from github...
When building from git I used the following commands:

sh ./autogen.sh --enable-maintainer-mode (no error but not sure if this was required)

sh ./configure --prefix=/usr/local --enable-shared --enable-parallel --with-pic --enable-maintainer-mode (the last flag is the most relevant here)

Until I did this meep would not compile because it would not build the missing scheme and python files, causing the build to fail. MANY ITERATIONS of make clean, make distclean, autogen.sh, configure, make were completed before this work.

**Comment by stevengj** [comment]
> If I try to build libctl --without-guile (I know that's a separate project) it fails at *** No rule to make target 'geom-ctl-io.c', needed by 'all'. Stop.

You need to use the official release tarball `libctl-4.5.0.tar.gz` (at https://github.com/NanoComp/libctl/releases/tag/v4.5.0) in order for `--without-guile` to work, not build from the git repo.  The tarball includes `geom-ctl-io.c`, whereas with the git repo you need to generate.

(Building from the raw git repo requires special software, e.g. `guile` and `automake`, to generate files like `geom-ctl-io.c`.)

**Comment by stevengj** [comment]
Yes, you need `--enable-maintainer-mode` to build from the git repo, as opposed to a release tarball.  (You don't need to pass this to `autogen.sh`, which already passes it to `configure`. )

**Comment by stevengj** [comment]
It should be possible to build Meep `--without-libctl` (though you then won't be able to use the Python or Guile APIs and can only use the lowest-level C++ API).   I'll check on this — it's possible that option has bitrotted.

**Comment by stevengj** [comment]
Whoops, no, it only succeeded with `--without-libctl` because I had libctl installed on my machine.  If I set my CPPFLAGS so that libctl is not discoverable then it fails to build.

**Comment by doswellf** [comment]
Thanks All!

After switching over to OpenSUSE 15.2 and installing guile and some 
other prerequisites I was able to configure and build meep with  scheme 
and python as well as open mpi and parallel hdf5 support. Haven't tested 
everything yet, but the configure and make worked. Currently updating 
the python docs with 'mkdocs serve' and will then run tests and install. 
Then make sure all the mpi-based cluster features work too.

Also currently building llvm, clang, openmp from source. Eventually I 
want to build a python package and a conda package that use parallel hd5 
and parallel meep over openmp. This meant the stock openSUSE 15.2 
packages were not useful. Also the current conda packages do not have 
parallel hdf5 enabled. I may install the latest slackware somewhere and 
build everything from source to eliminate package confusion.

Fraser
